inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/convolution/convolution_kernel_b_fs_zyx_fsv16_imad.cpp

```diff
@@ -63,190 +63,210 @@ static size_t getOutBlock_X(const size_t output_size_x, const size_t stride_x, c
+                            feature_slm_split = split;
+                        }
+                    }
+                }
+            }
+            if (split * fsv >= params.weights.IFM().v)
+                break;
         }
     }
 
     return BlockParams{ block_width, block_height, block_depth, block_features, in_block_width, in_block_height, in_block_depth, feature_slm_split };
 }
 
+float Convolution_kernel_b_fs_zyx_fsv16_imad::EstimateBlockParamsRatio(const convolution_params& params, const BlockParams& block) const {
+    float occupancy_by_logic_size = static_cast<float>(params.output.LogicalSize() / GetMaxThreadsPerDevice(params));
+    bool increase_max_reg_pressure = occupancy_by_logic_size >= 595.f;
+    bool twice_increase_max_reg_pressure = occupancy_by_logic_size >= 595.f * 2.f;
+    float max_reg_pressure = twice_increase_max_reg_pressure ? 0.785f : increase_max_reg_pressure ? 0.75f : 0.7f;
+
+    constexpr float max_occupancy = 2.f;
+    constexpr float max_slm_usage = 1.f;
+
+    // Estimate occupancy, slm usage and register pressure
+    float occupancy = EstimateOccupancy(params, block);
+    float slm_usage = EstimateSLMUsage(params, block);
+    float reg_pressure = EstimateRegPressure(params, block);
+
+    // Estimate fb32 usage factor
+    auto& output = params.output;
+    float feature_block_32 = static_cast<float>(block.output_block_features == 32);
+    float fb32_factor = -5.f;
+    if (params.engineInfo.deviceTypeIsDiscreteGPU && params.engineInfo.bIMADSupport) {
+        // Known cases where fb32 for discrete GPU works better
+        // many cases
+        bool fb32_exception_1 = output.X().v % 13 == 0 && output.X().v * output.Feature().v == 13312;
+        bool fb32_exception_2 = (output.X().v % 28 == 0 && output.X().v * output.Feature().v == 14336) || (output.X().v == 14 && output.Feature().v == 512);
+
+        // vehicle-attributes-recognition-barrier-0039
+        bool fb32_exception_3 = (output.X().v == 5 || output.X().v == 9) && output.Feature().v == 128;
+        bool fb32_exception_4 = output.X().v == 18 && output.Feature().v == 64;
+
+        // ctpn
+        bool fb32_exception_5 = output.X().v == 37 && output.Feature().v == 512;
+
+        // googlenet-v4
+        bool fb32_exception_6 = output.X().v == 17 && output.Feature().v == 256;
+
+        // all exceptions for z == 1
+        bool fb32_exceptions = fb32_exception_1 || fb32_exception_2 || fb32_exception_3 || fb32_exception_4 || fb32_exception_5 || fb32_exception_6;
+
+        // i3d-rgb
+        bool fb32_exception_z = output.X().v == output.Y().v && output.X().v % 28 == 0 && output.Z().v == 40 && output.Feature().v % 32 == 0;
+
+        if ((output.X().v == output.Y().v && output.Z().v == 1 && fb32_exceptions) || fb32_exception_z)
+            fb32_factor = 1.f;
+    } else if (occupancy_by_logic_size >= 2500.f) {
+        fb32_factor = 0.5f;
+    }
+
+    float reg_pressure_factor = atanf(occupancy) / 3.14159f;
```

Comment:
Could you add some comments with explanation of these heuristics? Many of these formulas are far from being obvious...

Reply:
Arctangens is a symmetric function with positive values for x > 0 (we are only interested in positive values because the occupancy is always a positive value). For low occupancy (e.g. < 1) we shouldn't concentrate our attention on reg_pressure_factor because target #1 in these cases is the occupancy increase. So for occupancy < 1 reg_pressure factor is enough low and it's growing with occupancy's grow. Pi number (3.14159f) is a scaling coefficient for setting function value in range [0, 0,5f].

Reply:
great, thanks. Now copy it to the code :)